Changes to mitigate white flashes in heavier themes

pjft opened the pull request:
Hi. I'm proposing to add two options that will help improve the perceived performance in heavier themes.

- Option to not use multithreaded image loading, like old ES (default: use multithread, like today)
- Option to re-use SVG in-memory images in same path (default: not re-use, like today)

In the recent WSOD changes, one improvement was to load images in separate threads. While this works well for non-heavy themes, it causes an unpleasant white screen to show before the actual images are rendered, when they need to be cycled in the texture manager memory. Forcing them to be loaded like the old ES, in the same thread, will cause a potentially less smooth system view scrolling, at the exchange of no more white screens.

Also, SVGs from the same path are not re-used throughout the theme, but actually re-created in memory. In particular, borders and background SVGs take up a significant amount of memory. Adding that option will allow a significant reduction in memory consumption.
I'm not making it default because in the code it stated that it was disabled for SVGs because the same SVG can be rasterized at different sizes, and I imagine this change will just keep the first texture it loads, and then resize in runtime as needed, which - if upscaling - may lead to lower quality if effectively re-used in different sizes. 
Tested Carbon and it rendered just fine - I assume it either loads first the larger-sizes controller SVGs, so in the gamelist view it just resizes them to be smaller - or that it's not an issue, as Carbon does re-use the controller SVGs across system and gamelist views.

pjft commented:
Also set the default GL clear color to black rather than the previous white.

pjft commented:
... Thinking about it a bit more, now that the default color is black, the option for single threaded image loading is kind of redundant as both will behave quite similarly but the multi threaded option does not offer any significant advantage, and adds clunkiness to the mix. 

Would love other people's thoughts but I'm open to scraping that option and change altogether unless someone feels there's a case for keeping it, but I'd rather not add yet another option to the mix if it doesn't add much benefit.

HerbFargus commented:
>especially as the current transitions fade from black

Is this the same fade you're referring to? I thought it was white? 

https://github.com/RetroPie/EmulationStation/pull/52

Or are you saying you were able to change the fade to Black? If so my black splash may work with this change

pjft commented:
Unsure if we're talking about the same thing :/

I got some videos to demonstrate:

https://www.youtube.com/watch?v=A9yGEm1MCAY&list=PLyOevcafAnDSp-eH24CIv2sjOP0QQt9ru

Apologies for the baby sounds in the background :)

The white ones are from the current build. Then I go through single/multi
thread, and fade/slide transitions.

Hopefully it'll help to understand what this helps with.


- Paulo Tavares

On Sat, May 13, 2017 at 1:40 PM, HerbFargus <notifications@github.com>
wrote:

> especially as the current transitions fade from black
>
> Is this the same fade you're referring to? I thought it was white?
>
> #52 <https://github.com/RetroPie/EmulationStation/pull/52>
>
> Or are you saying you were able to change the fade to Black? If so my
> black splash may work with this change
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RetroPie/EmulationStation/pull/129#issuecomment-301245768>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVAV7Zcsq8odbgNjXSHF6Jv59f6RS09Yks5r5aSzgaJpZM4NZ2Mf>
> .
>

HerbFargus commented:
I really like the black.

When you boot up the first time does it fade to Black now instead of fading to white? If I can find time I'll see if I can give it a test myself but just from the videos I definitely am preferable to the black

pjft commented:
I think so, but I can check. Since I have a random video at startup it usually ends when it's already loaded. You're referring to the fade after the ES splash, is that it? Is that something you were looking to change in the other PR? I lacked context there, but happy to help.

The white is the current version, but truth be told I think it looks worse than the black, which is the default "blank" color in the screens we use. The GL window was initialized as white, so changing that to black already makes it look "cleaner". The white flashes look like an error, while black just look like it may be taking some more time if needed.

HerbFargus commented:
Yeah in the other PR I'm pretty sure the opening fade either right before the splash loads or right after is the same fade that is used in the fade transition so if you've managed to turn it black then my other PR should now be viable. Would match the default carbon theme better too

pjft commented:
I'll see what's happening on my builds and get a video but I never really
noticed much. If you have a video on your end that can help me best
understand what you're experiencing and I'm happy to take a look.
On Sat, 13 May 2017 at 17:14, HerbFargus <notifications@github.com> wrote:

> Yeah in the other PR I'm pretty sure the opening fade either right before
> the splash loads or right after is the same fade that is used in the fade
> transition so if you've managed to turn it black then my other PR should
> now be viable. Would match the default carbon theme better too
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RetroPie/EmulationStation/pull/129#issuecomment-301258039>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVAV7bSr7hEcMFa79GoKVcHybuBuEyZzks5r5db4gaJpZM4NZ2Mf>
> .
>

HerbFargus commented:
Yeah no worries. My free time has been limited as of recent but if/when I get a moment I'll test out your changes with my other PR to see if it's what I think it is. I have a feeling it is, and then hopefully we can get some feedback from some others as well.

pjft commented:
Removed the options from the menu, and opted to re-use the SVGs by default on the Pi, especially after seeing the report from 

https://retropie.org.uk/forum/topic/10197/white-flicker-on-moving-through-ui-after-update-to-4-2-3/26

on Carbon, which is very easily reproducible, and stems from the fact that the logos on Carbon are SVGs, which are not well managed in memory at the moment.

Summarizing the changes from the PR, then:
- Default GL blank color changed from white to black;
- Re-using SVGs textures from the same file path in memory, rather than creating one instance for each, to reduce/remove flicker;
- Added setting (internal, not in menu) to force images to be loaded in single-thread. I'm happy to remove this altogether, or add it to the menu, though - if I understand correctly - one can still set this in the settings file manually if needs be and it seems not to be needed unless one really doesn't want to see any flicker. 

Thoughts are welcome. I'll squash it when/if we're comfortable with the changes.

pjft commented:
Update: Reverted the last change with the defaults. The SVG option is useful in several cases, particularly in themes that don't re-use the same SVG in different sizes (such as Carbon) - but can't be a default, at least not yet, precisely because of that.

If the SVG is rendered in a lower res first (as it happens on leaving a game), the higher-res SVGs will be rendered in lower resolution. I'll see if I can adjust the code to make it work best (i.e. replace the SVG if rendering at a higher res), or see if keeping the options would be a good approach, for people to adjust should their themes not perform well?

More to come.

HerbFargus commented:
I gave it a bit of a test but I'm not a romhoarder so I wouldn't know how it performs with massive romsets. But it seemed to function well enough and I tested out my splash screen changes too and those also work without the fade issues so I am satisfied. Would probably be good to get others input who have more systems.

joolswills commented:
I've been a bit busy so wasn't able to comment. I noticed the flashing on Carbon, with not so many systems. It happens when entering a system with metadata/box pics, then returning to the system list. the images that are already on screen flash back to white, then back to the logo. I assume this is because they have been cleared from ram to make space for the meta data gfx etc, but I guess it could be a bug.

I intended to look into/bisect this, just to check/confirm when the issue started.

Sorry if this is known/not useful. I haven't had time to read all the stuff about this.

pjft commented:
Thanks all for the comments.

@HerbFargus Thanks for testing - I'm happy it works for you, as well as enabling you to sort you the splash screen changes! We ran into some troubles yesterday after testing - re-using SVGs is fine, but if the lower res SVG was loaded first, the others would be upscaled which was sub-optimal (and that's an euphemism if I ever heard one).

I fixed it, and removed the settings from the menus as well as the option to do single-thread loading. It isn't necessary for now, unless someone really feels the need.

The latest changes (separate commit, but can squash after testing) will ensure that we always keep the highest rasterized resolution of a SVG, given that in runtime we will always scale it (down- or up-scale) as needed via the ImageComponent UI component. This way we save memory by only storing one in-memory instance of each file, and avoid the flashing issues in Carbon, which - as you tested, @joolswills , are definitely related to the SVGs being kicked out of memory, but also not being well managed. In particular, it was quite egregious when we launched a game (which unloads all textures), and then come back (which reloads them), since it would not have the necessary versions loaded at the time.

This should definitely help - do test it out.

I'll be testing a bit more as well, but hopefully the changes as they stand are simple enough to be approved:
- Default black color as "blank" for GL canvas;
- Default for RPI to re-use the SVGs in memory for the same path, keeping the highest rasterized version only in memory, and scaling appropriately via ImageComponent.

@HerbFargus - I believe you use Windows as well, correct? Can you test the Carbon issues on Windows and see whether they happen there? I expect it to have a lot more video memory so hopefully that won't be an issue, but if it is I can make re-using SVGs a default for all systems rather than just the Pi.

pjft commented:
@joolswills It seems I've not reached any particularly satisfactory conclusion on my end, as while the issue has been reduced, it is still there.

I'll keep trying to track this one down in the coming days/weeks, time permitting.

At the moment, the PR is still not complete.

TMNTturtleguy commented:
@pjft  I have done extensive testing of your update, I left comments on the forum thread as well.  The proposed update is awesome.  My testing did show there is still work to be done, but I feel it is a major upgrade to the current White Screen Of Death Fix, even if not perfect.  Below I will detail my testing and provide videos of what I am seeing.  At this point i have not gone back to record video of how this performed prior to this update. If @HerbFargus and @joolswills would like those "before" videos I can reinstall the base ES and make them.  I also want to note that I could create white screens of death on the main ES build as well, so the WSOD is not necessarily a symptom created by these modifications.

 Test 1: Comic Book Theme (heavy theme, 17 systems loaded, uses .svg files) VRAM at 100
- After boot up it looks amazing, i can slide with little to no stutter at all, the slide is very smooth, logos and backgrounds load very fast and clean.  There are some black screens between some systems loading, but compared to the very large white flashing, this looks so smooth and clean.
- After entering a game and exiting: .png background reload time can vary, sometimes fast, sometimes a slight stutter. Exiting back to system select screen.  The first time scrolling to the right, there is a large black flash, and the .png system logos now have a white box flash for a half second before the .png logo loads.  After one full cycle, everything goes back to completely smooth.  I also noted this does not happen after every time i exit a game. Sometimes, it works smoothly.

Video of this test: [youtube Retropie ES white flash fix test ](https://youtu.be/Y2PUJNYanWs)
- After continued testing I have been able to create following errors only after entering into the video view (rom selections screen), @ VRAM 120 - If I scroll super fast through several systems over and over, i can create glget errors.  It also can create the meta data images in the video view (game select view) to turn white.  The main background is there, but the meta data images are white boxes the size of the max size regions set.  @ VRAM 100 - I can recreate the glget error after continuous super fast scrolling.  Have not been able to recreate the meta data loss.  @ VRAM 110 - I can recreate both glget error and white meta data, but it is harder to recreate this.  Video Link [youtube glget error](https://youtu.be/YBt94M7eoEs)

Test 2: Carbon Theme 17 systems loaded VRAM 100
- This reacts the exact same as the comic book theme did.  I notice the same stutter after exiting roms and the same white box on the logos when sliding right.  These logos are .svg in carbon.  From my testing, it seems like your current solution as fixed the use of .svg and .png, not it is just figuring out how to get the white boxes to disappear after exiting a game.
- Using Carbon I cannot create any glget errors.  I do the same test as above by entering the video view and existing to system select view and scroll very fast, i cannot create a glget error.  I believe this proves that the heavy themes that may have had issues could still have issues after the update, and the update itself is not causing those issues.  It should be noted that prior to this update, it was impossible to scroll that quickly in the comic book theme, and when scrolling fast with 17 systems i could produce glget errors on the main branch build.

 Test 3: Comic Book Theme (heavy theme, 17 systems loaded, uses .svg files) VRAM at 120
- This is absolutely amazing!  I can scroll through as fast as I can hit my arrow buttons and never flash or loose the background!  After existing a game, the first time though there is a black flash and the logos are still white, after the first time through, it is back to lightning fast speeds.

Video of this test: [youtube White flicker test VRAM 120](https://youtu.be/rvHJW9CfJho)

- See above under test 1 for errors found while using VRAM 100, 110, and 120.

I also tested this on Futura, Workbench, and Tronkyfran and had great success and did not experience the glget after exiting the video view screen.  I believe I have identified the fact that the comic book theme has 2 .svg files, 3 meta data images, and 1 .png image loaded on top of the .png background.  No other theme tested has that many items loaded into the screen.   I also want to note that I don't know why anyone would continuously scroll that fast through screens unless they were trying to break it!  (which I was).  I believe this update is a major step in the right direction.  This is a major performance upgrade over the current WSOD fix with the exception of the white boxes over the system logos.  The performance upgrade is worth the white box over the logo as we were all getting the white flashes before, this limits the flash to the logo.  Hopefully a fix can be found for that as well moving forward.  I plan to keep this as my primary ES build and will provide more updates if I find anything.

pjft commented:
Ok, so a few interesting findings, for better and for worse:
- I ran an old build with the WSOD fix, and indeed this didn't happen.
- I also ran a development build I had from back in early April (which, unfortunately, doesn't mean much as I was on a separate branch), and it didn't happen as well. It seems to be based on 2.1.4, though, so that could be a starting point.
- I ran a build from early May, and it was certainly present.

I just read about git bisect. Looks... err, fun, for lack of a better word.

I may try it out to try and track this down.

Thanks for the pointer.

joolswills commented:
I suspected it was since the original WSOD fix, as I never noticed it then. I'm happy to bisect it later today.

TMNTturtleguy commented:
For what it's worth, based on my memory (so it probably isn't worth much), I seem to remember a good upgrade with the WSOD fix, however there was the issue with dark boxes over meta data fields.  I believe i remember seeing issues arise after the update by jdrassa which fixed the black box issue.  It could just be that my eye was caught by the black box and I missed these other issues.

pjft commented:
Thanks all.

@joolswills I'm more than happy to try it out myself, it should prove a useful learning experience. Worst case, it'll build character :)

I appreciate the offer for help, certainly, and if I do end up getting stuck or running into a dead end I'll certainly come back and take you up on that, but I've seen that you have plenty on your plate already between the forums, the project, and the other requests coming from all over the place!

Cheers.

@TMNTturtleguy I was assuming that the WSOD fix would have marked the start of it as well, but you might be right. At least from my earlier test, I do believe I tested a build post-WSOD fix and it seemed to not be flashing after loading and exiting a game, so I'm intrigued.

For what it's worth, I'm starting the tests with this commit, which marks the introduction of the WSOD fix (well, the version update post-WSOD fix).

https://github.com/RetroPie/EmulationStation/commits/master?after=268e6c8338121c735cd607d5280ee36c8077fbcb+34

I'll get back to you all on this.

Thanks!

pjft commented:
Yay - or halfway there!

I was able to trace it to this: 

 @zigurana 

```
ac5e3ad95f34ac52bc925d618f37fe9b1717b42b is the first bad commit
commit ac5e3ad95f34ac52bc925d618f37fe9b1717b42b
Author: D. Polders <xxx@xxx.xxx>
Date:   Mon Mar 13 22:11:07 2017 +0100

    System Carousel feature, now with only a single commit!

:100644 100644 685ee8c42674f522e074144c39c63b706de18056 0686c4180e8b19add800f75203bcbc1c13776869 M	.gitignore
:100644 100644 d15b4303eb2c3b46c397c54d8cc78ded114c7e24 1b98f162169a6e2326dccf1ccf7c89c64734f6a8 M	CMakeLists.txt
:100644 100644 a8a111fa13a0c7b7a6e2e841ae2428b85d9084fe 1c518dafd58de536f0cc8253ed39f82d53ff9971 M	THEMES.md
:040000 040000 7da44b5e4bd7e14493f8842c25049d7e9313b4ba 47c2368696bb60dbe7b972893ca1233119e6b435 M	es-app
:040000 040000 adcf6140b7f0a8ce5ee6288d02fb99abcab86fb7 956ff98b0f318308f893daedea59faca0527e1dc M	es-core
```

I was checking that and it's just a mammoth commit. I'm going to cry myself to sleep in the corner now.

:)

@joolswills if there's any finer-grained insights one can derive from bisect and such, do let me know - but even in the original repository there's only a squashed commit and no similar branches.

I'll go dig into it later - it does make sense that, given the huge work put into it and the large refactoring done to some of those, that something could have gone awry. It's a testament to how solid it was that it managed to go by for so long until anything even remotely cropped up.

Zigurana: if you have any idea of what might be causing textures now to flash when loading, especially after loading and exiting from a game (i.e. after being reloaded) now is the time to speak up, as it could save me a bit of digging around :)

Thanks all for the pointers!

pjft commented:
Ok, figured it out.

Before I implement the fix, though, I'd love to hear from @zigurana if there's anything that you anticipate these will break if I revert these two changes (there are two lines with it):

https://github.com/RetroPie/EmulationStation/commit/ac5e3ad95f34ac52bc925d618f37fe9b1717b42b#commitcomment-22145345

Thanks!

zigurana commented:
Man, cut it out with all these emails! I am trying to navigate github on my
mobile, and I cant keep up with you! :-D
Give me a moment, and I'll try to have a look.

Op ma 15 mei 2017 18:41 schreef pjft <notifications@github.com>:

> Ok, figured it out.
>
> Before I implement the fix, though, I'd love to hear from @zigurana
> <https://github.com/zigurana> if there's anything that you anticipate
> these will break if I revert these two changes (there are two lines with
> it):
>
> ac5e3ad#commitcomment-22145345
> <https://github.com/RetroPie/EmulationStation/commit/ac5e3ad95f34ac52bc925d618f37fe9b1717b42b#commitcomment-22145345>
>
> Thanks!
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/RetroPie/EmulationStation/pull/129#issuecomment-301532129>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF0i2OH73M54jCzMFGClj3OzZQ6T3UWpks5r6IBKgaJpZM4NZ2Mf>
> .
>

zigurana commented:
Hey!
Damn that's a good catch! I can't recall, which means that its probably an
inadvertent ommission. Probably a merge error while updating my branch with
the WSOD changes.

Sorry for not testing this better...

Op ma 15 mei 2017 18:47 schreef Zigurana <zigurana@gmail.com>:

> Man, cut it out with all these emails! I am trying to navigate github on
> my mobile, and I cant keep up with you! :-D
> Give me a moment, and I'll try to have a look.
>
> Op ma 15 mei 2017 18:41 schreef pjft <notifications@github.com>:
>
>> Ok, figured it out.
>>
>> Before I implement the fix, though, I'd love to hear from @zigurana
>> <https://github.com/zigurana> if there's anything that you anticipate
>> these will break if I revert these two changes (there are two lines with
>> it):
>>
>> ac5e3ad#commitcomment-22145345
>> <https://github.com/RetroPie/EmulationStation/commit/ac5e3ad95f34ac52bc925d618f37fe9b1717b42b#commitcomment-22145345>
>>
>> Thanks!
>>
>> —
>> You are receiving this because you were mentioned.
>>
>>
>> Reply to this email directly, view it on GitHub
>> <https://github.com/RetroPie/EmulationStation/pull/129#issuecomment-301532129>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AF0i2OH73M54jCzMFGClj3OzZQ6T3UWpks5r6IBKgaJpZM4NZ2Mf>
>> .
>>
>

joolswills commented:
Not need to apologise :-) Appreciate your improvement/contributions. Thanks @pjft for debugging this.

pjft commented:
Thanks all, and thanks @zigurana for the quick reply. I reverted the change as well at the moment, though I'm still not fully happy with the final outcome. 

Reverting the changes fixes things for Carbon, but brings it back to where we were before, for heavier themes, with the white flashing icons. I re-added the "Re-Use SVGs" option to the menu, as it can't be a default at the moment - for some reason, it causes some flashing on the Carbon SVG logos after exiting from a game, so I am going to look into that.

If I can solve that and make re-using SVGs the default, as it saves a lot of memory, I'll be happy with the outcome. 

@zigurana I apologize for cluttering your inbox - I'm not sure if your comment was related to just yesterday or in general, but I do appreciate everyone's guidance in terms of GitHub updates best practices. :)

I do acknowledge that I end up adding a lot of "developer logs" here, mostly so others who may be reviewing or testing the code can know the exact status and problems I'm facing (if any) so that they may help or at least know what to expect. I can do it by editing the previous post, though I would usually only do it if not a lot of time has elapsed since I originally posted.

Guidance appreciated, definitely - and thanks for the feedback!

zigurana commented:
Hey man, don't worry, I was fumbling around on my phone trying to find the right line of code, while your messages were pouring in. Not an issue, I'll cope :-).
Thanks for picking this up, you are doing great work for the project!

joolswills commented:
Is the alsa change related to this ? If not, please can you do that as a separate PR.

joolswills commented:
Sorry ,I misread - I see it is related.

pjft commented:
Ok, I'm happy with this PR now. After diving a lot back and forth into the code, memory management, etc, this is the final summary of the changes:

- Set default GL clear color to black, rather than white, for canvas initialization white flashes
- Revert changes to ImageComponent intialization that cause white flashes on logos
- Increased ALSA buffer to prevent buffer underruns which stall CPU momentarily, causing textures to not load on time
- Improved carousel texture buffering based on navigation direction and speed

I'm happy that in the end there are no new options to be added, but rather making the overall performance better for everything.

I know it's a bit of a mixed bag, so happy to split some of these (namely the ALSA one) if preferred. I have no significant objections.

Thanks.

zigurana commented:
Thanks for the synopsis, I had difficulty keeping track of what changes
were still relevant to fix this.
Thank you for your time!

Op wo 17 mei 2017 21:56 schreef pjft <notifications@github.com>:

> Ok, I'm happy with this PR now. After diving a lot back and forth into the
> code, memory management, etc, this is the final summary of the changes:
>
>    - Set default GL clear color to black, rather than white, for canvas
>    initialization white flashes
>    - Revert changes to ImageComponent intialization that cause white
>    flashes on logos
>    - Increased ALSA buffer to prevent buffer underruns which stall CPU
>    momentarily, causing textures to not load on time
>    - Improved carousel texture buffering based on navigation direction
>    and speed
>
> I'm happy that in the end there are no new options to be added, but rather
> making the overall performance better for everything.
>
> I know it's a bit of a mixed bag, so happy to split some of these (namely
> the ALSA one) if preferred. I have no significant objections.
>
> Thanks.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/RetroPie/EmulationStation/pull/129#issuecomment-302214228>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF0i2NzPUDGyG3fWVUNNPaeL7qElxxsaks5r61D5gaJpZM4NZ2Mf>
> .
>

pjft commented:
"I had difficulty keeping track of what changes were still relevant to fix this".

You probably imagine I had similar difficulties, by the amount of back and forth here :)

Thanks for putting up with this.

joolswills commented:
Thanks very much for this. I'm happy for all in one, but if you wanted to split it into a few commits feel free.

pjft commented:
I already kind of squashed it on my end, and am happy with that - there aren't that many files changed, so I'll leave it up to you. 

There are some folks testing it on the forums - TMNTTurtleGuy and mediamogul have said it works well, which is reassuring - so let's see how it goes.

And I learned quite a few things so, as unstructured as it was, I'm quite happy with having learned to bisect in Git, to understand more about the texture pipeline in ES, the SystemView rendering, and to have improved it overall for everyone. So thanks for that!

joolswills commented:
thanks.
